logstore/lstoreds/headbook.go

```diff
@@ -157,20 +161,28 @@ func (hb *dsHeadBook) ClearHeads(t thread.ID, p peer.ID) error {
 }
 
 func (hb *dsHeadBook) HeadsEdge(tid thread.ID) (uint64, error) {
+	var key = dsThreadKey(tid, hbEdge)
+	// computing and writing back previously invalidated
+	// edge frequently results in transaction conflicts
+	for attempt := 1; attempt <= 3; attempt++ {
```

Comment:
large instances would better use hybrid logstore with edges fetched from the in-memory storage to avoid performance issues

Reply:
As the reason of the issue is contention on the thread's head-keys, I believe that optimal solution depends on the intensity of writes to the logs. While contention is relatively low, I'd rather use optimistic approach with retried transactions. In case of unusually intensive writes and high contention pessimistic locking/semaphores would perform better.